metadataserver/cni/cni_linux.go

```diff
@@ -0,0 +1,420 @@
+	"golang.org/x/sys/unix"
+
+	"github.com/sirupsen/logrus"
+	"github.com/wercker/journalhook"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+
+	"github.com/Netflix/titus-executor/utils"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+)
+
+const (
+	contName = "imdsproxy"
+	peerName = "container"
+)
+
+func init() {
+	// namespaces are per task
+	runtime.LockOSThread()
+
+	// use journald for logs
+	journalhook.Enable()
+}
+
+func getPeerNsPath(podName string) string {
+	return fmt.Sprintf("/var/run/netns/imds-%s", podName)
+}
+
+func mountNs(ns netns.NsHandle, peerNsPath string) error {
+	err := netns.Set(ns)
+	if err != nil {
+		return errors.Wrap(err, "netns.Set")
+	}
+
+	mount, err := os.Create(peerNsPath)
+	if err != nil {
+		return errors.Wrap(err, "os.Create")
+	}
+	_ = mount.Close()
+
+	err = unix.Mount(utils.CurrThreadNetNs, peerNsPath, "none", unix.MS_BIND, "")
+	if err != nil {
+		return errors.Wrap(err, "unix.Mount")
+	}
+
+	return nil
+}
+
+func getContainerIP(prevResult *current.Result) (net.IP, error) {
+	for idx, iface := range prevResult.Interfaces {
+		if iface.Name == "eth0" {
+			return prevResult.IPs[idx].Address.IP, nil
+		}
+	}
+	return nil, errors.New("could not find default interface")
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	logrus.Debugf("add %#v", args)
+
+	prevResult, err := getPrev(args)
+	if err != nil {
+		logrus.Errorf("getPrev %s", err)
+		return err
+	}
+	logrus.Debugf("getPrev %#v", prevResult)
+
+	containerIP, err := getContainerIP(prevResult)
+	if err != nil {
+		logrus.Errorf("getContainerIP %s", err)
+		return err
+	}
+
+	pod, err := getPod(args)
+	if err != nil {
+		logrus.Errorf("getPod %s", err)
+		return err
+	}
+	logrus.Debugf("getPod %#v", pod)
+
+	env, err := extractEnv(prevResult, pod)
+	if err != nil {
+		logrus.Errorf("extractEnv %s", err)
+		return err
+	}
+	logrus.Debugf("extractEnv %#v", env)
+
+	// create new namespaces
+
+	peerNs, err := netns.New()
+	if err != nil {
+		logrus.Errorf("peerNs %s", err)
+		return err
+
+	}
+	logrus.Debugf("peerNs %#v", peerNs)
+
+	contNs, err := netns.GetFromPath(args.Netns)
+	if err != nil {
+		logrus.Errorf("contNs %s", err)
+		return err
+	}
+	logrus.Debugf("contNs %#v", contNs)
+
+	// create veth from container ns
+
+	veth := &netlink.Veth{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:      contName,
+			Namespace: netlink.NsFd(contNs),
+			Flags:     net.FlagUp,
+		},
+		PeerName: peerName,
+	}
+	logrus.Debugf("veth %#v", veth)
+
+	if err := netlink.LinkAdd(veth); err != nil {
+		logrus.Errorf("LinkAdd %s", err)
+		return err
+	}
+	logrus.Debugf("LinkAdd %#v", veth)
+
+	// forward declare for gotos
+
+	var peerLink, contLink netlink.Link
+	var route *netlink.Route
+
+	var peerNsPath string
+	var errA error
+
+	// move the peer veth into the peer namespace, link up
```

Comment:
Thank-you for all of the high-level comments in this file - made following it a lot easier!